scripts/abundance-dist-single-threaded.py

```diff
@@ -0,0 +1,254 @@
+                        action='store_true')
+    return parser
+
+
+def main():  # pylint: disable=too-many-locals,too-many-branches
+    args = sanitize_help(get_parser()).parse_args()
+
+    configure_logging(args.quiet)
+    report_on_config(args)
+
+    check_input_files(args.input_sequence_filename, args.force)
+    if args.savegraph:
+        graphsize = calculate_graphsize(args, 'countgraph')
+        check_space_for_graph(args.savegraph, graphsize, args.force)
+    if (not args.squash_output and
+            os.path.exists(args.output_histogram_filename)):
+        log_error('ERROR: {output} exists; not squashing.',
+                  output=args.output_histogram_filename)
+        sys.exit(1)
+    else:
+        hist_fp = open(args.output_histogram_filename, 'w')
+        hist_fp_csv = csv.writer(hist_fp)
+        # write headers:
+        hist_fp_csv.writerow(['abundance', 'count', 'cumulative',
+                              'cumulative_fraction'])
+
+    log_info('making countgraph')
+    countgraph = khmer_args.create_countgraph(args, multiplier=1.1)
+    countgraph.set_use_bigcount(args.bigcount)
+
+    log_info('building k-mer tracking graph')
+    tracking = khmer_args.create_nodegraph(args, multiplier=1.1)
+
+    log_info('kmer_size: {ksize}', ksize=countgraph.ksize())
+    log_info('k-mer countgraph sizes: {sizes}', sizes=countgraph.hashsizes())
+    log_info('outputting to {output}', output=args.output_histogram_filename)
+
+    # start loading
+    reads_queue = Queue(maxsize=1000)
+    n_consumers = 1
+
+    def _load_reads(fname, q, n_consumers):
+        rparser = khmer.ReadParser(fname)
+        chunk = []
+        for n, read in enumerate(rparser):
+            chunk.append(read.sequence)
+            if not n % 1000:
+                q.put(chunk)
+                chunk = []
+        if chunk:
+            q.put(chunk)
+        # Each consumer must only consume one sentinel otherwise things
+        # will hang.
+        for _ in range(n_consumers):
+            q.put(None)
+
+    def _consume_reads(cg, q):
+        while True:
+            reads = q.get()
+            if reads is None:
+                break
+            cg.consume_chunk(reads)
+
+    loading_thread = threading.Thread(target=_load_reads,
+                                      args=(args.input_sequence_filename,
+                                            reads_queue,
+                                            n_consumers))
+    loading_thread.start()
+
+    consumers = [threading.Thread(target=_consume_reads,
+                                  args=(countgraph, reads_queue))
+                 for n in range(n_consumers)]
+    _ = [t.start() for t in consumers]
+
+    log_info('consuming input, round 1 -- {input}',
+             input=args.input_sequence_filename)
+
+    for thread in consumers:
+        thread.join()
+    loading_thread.join()
+
+    log_info('Total number of unique k-mers: {nk}',
+             nk=countgraph.n_unique_kmers())
```

Comment:
Looks familiar to me :) -- see [https://github.com/dib-lab/khmer/blob/master/khmer/thread_utils.py#L73](this), which @camillescott once convinced me didn't add anything to the speed of things.

So I like the pattern, if it can be made fast!